### PR TITLE
[DATA-2728] close datamanger if reconfigure fails and tolerate booting offline

### PR DESCRIPTION
**Pull request by nicksanford**
Bandaid fix for: https://viam.atlassian.net/browse/DATA-2728?focusedCommentId=31943

1. Change datamanager to call Close() on itself if Reconfigure returns an error
2. Change datasync to config changes to be processed after Reconfigure so that datamanger can boot successfully if viam-server starts while offline

### Manual Testing:

I tested this with a custom sensor I wrote that records the number of times Readings has been called & the current time when the Readings method was called. This allowed me to confirm that data capture captured data at the configured interval (once a second) and that no sensor readings were dropped.

#### Test Cases

1. Booted a Pi with a viam-server built from this PR with an internet connection & the config below. Collected data for a few minutes. Unplugged pi from internet. Let run for a few minutes. Re connected pi To internet. Let viam-server run for a few more minutes. Stopped viam-server. Ran export command to collect data from when viam-server was running. Confirmed that no readings were missed & data capture captured data every second.
2. Booted a Pi with a viam-server built from this PR WITHOUT an internet connection & the config below. Collected data for a few minutes.  Re connected pi from internet. Let viam-server run for a few more minutes. Unplugged Pi from internet. Let run for a few minutes. Reconnected pi to the internet. Let data flush to teh cloud. Stopped viam-server. Ran export command to collect data from when viam-server was running. Confirmed that no readings were missed & data capture captured data every second.
3. Changed config values while robot is online & confirm config changes propagate

#### manual testing config
```json
{
  "components": [
    {
      "name": "boom",
      "namespace": "rdk",
      "type": "sensor",
      "model": "ncs:sensor:boom",
      "attributes": {},
      "service_configs": [
        {
          "type": "data_manager",
          "attributes": {
            "capture_methods": [
              {
                "method": "Readings",
                "capture_frequency_hz": 1,
                "additional_params": {}
              }
            ]
          }
        }
      ]
    }
  ],
  "services": [
    {
      "name": "data_manager-1",
      "namespace": "rdk",
      "type": "data_manager",
      "attributes": {
        "additional_sync_paths": [],
        "sync_disabled": false,
        "maximum_num_sync_threads": 10,
        "sync_interval_mins": 0.01,
        "capture_dir": "",
        "tags": [
          "nick"
        ]
      }
    }
  ],
  "modules": [
    {
      "type": "local",
      "name": "boomsensor",
      "executable_path": "/home/user/boomsensor"
    }
  ]
}
```
where `boomsensor` is a module produced from: https://github.com/nicksanford/boomsensor

### CODE REVIEW

File: services/datamanager/builtin/builtin.go
Comment (nicksanford) [inline services/datamanager/builtin/builtin.go:770]: remove

File: services/datamanager/builtin/builtin.go
Comment (nicksanford) [inline services/datamanager/builtin/builtin.go:411]: Previously if a user used a local config data manager just wouldn't initialize if they configured data sync. This changes it so that datamanger will initialize with local configs, data sync just won't do anything.

File: services/datamanager/builtin/builtin.go
Comment (nicksanford) [inline services/datamanager/builtin/builtin.go:428]: `syncerConstructor` never returned an errors so I made the signature reflect that.

File: services/datamanager/builtin/builtin.go
Comment (nicksanford) [inline services/datamanager/builtin/builtin.go:625]: This code has been moved into `asyncInitDataSync` so that it can run on an interval.

File: services/datamanager/builtin/builtin.go
Comment (nicksanford) [inline services/datamanager/builtin/builtin.go:460]: We call Close() on datamanager if Reconfigure returns an error (so that we terminate any goroutines started up until that point by Reconfigure.

File: services/datamanager/builtin/builtin.go
Comment (nicksanford) [inline services/datamanager/builtin/builtin.go:649]: when `svc.syncConfigUpdated == true` we know that a previous `Reconfigure`  has occured which changed datasync related config, which requires us to call initSyncer and or closSyncer.

Once we have processed that config this goroutine sets `svc.syncConfigUpdated` to false.

We assume that we don't need to successfully process ALL datasync related configurations, we just need to attempt to process the current one until either Close() is called on data manager, we determine this is a local config managed robot (which doesn't support datasync), or datasync is initialized with the new config.

File: services/datamanager/builtin/capture_test.go
Comment (nicksanford) [inline services/datamanager/builtin/capture_test.go:143]: In order to not rewrite the tests, I call `propagateDataSyncConfig` after every reconfigure which will happen at latest 1 second after the previous reconfigure in production.

File: services/datamanager/builtin/builtin.go
Comment (agreenb) [reply services/datamanager/builtin/builtin.go:411]: Would it throw an error here when it was a local config? Or below when calling svc.syncerConstructor? I would have expected that this would return a Noop syncer and thus still continue with a valid data manager service that captures.

**Comment by agreenb** [comment]
Included @n0nick as an fyi.

File: services/datamanager/builtin/builtin.go
Comment (n0nick) [inline services/datamanager/builtin/builtin.go:631]: nit, this code doesn't only check, it also deletes files in that case

```suggestion
	// if datacapture is enabled, kick off a go routine to handle disk space filling due to
	// cached datacapture files
```

File: services/datamanager/builtin/builtin.go
Comment (n0nick) [inline services/datamanager/builtin/builtin.go:670]: could you instead short-circuit this function?

```go
if !svg.syncConfigUpdated {
  return nil
}
// ...
```

File: services/datamanager/builtin/builtin.go
Comment (n0nick) [inline services/datamanager/builtin/builtin.go:692]: is this the same as L674-681 with the addition of `svc.CloseSyncer`? could we just DRY and have that close in a small `if` condition?

File: services/datamanager/builtin/builtin.go
Comment (n0nick) [inline services/datamanager/builtin/builtin.go:154]: nit, can this be renamed? I'm not sure what this represents as-is. something like, "syncer needs to be re-initialized"?

**Review by n0nick — COMMENTED** [review]
I trust your logic and tests, only had some style / clarificaiton comments

File: services/datamanager/builtin/builtin.go
Comment (dmhilly) [inline services/datamanager/builtin/builtin.go:653]: Just to confirm: the reason this could happen is if the robot was offline when we tried to propagate these changes last time?

**Review by dmhilly — COMMENTED** [review]
Left a couple questions; thanks Nick!

File: services/datamanager/builtin/builtin.go
Comment (nicksanford) [reply services/datamanager/builtin/builtin.go:411]: Previously if you booted viam-server with a local config that used data sync it would return an error here: https://github.com/viamrobotics/rdk/pull/4164/files#diff-ffb86a3e2b364cc470f33937bb1ffb08ca2ef9d01b2734320daea7bee24a8943L416 which would result in the Reconfigure method returning an error which would result in data management never booting.

Comment (nicksanford) [reply services/datamanager/builtin/builtin.go:411]: The fact that we set `svc.syncer = datasync.NewNoopManager()` doesn't change the fact that `initSyncer` returns an error.

File: services/datamanager/builtin/builtin.go
Comment (nicksanford) [reply services/datamanager/builtin/builtin.go:459]: Dan wrote it :)

File: services/datamanager/builtin/builtin.go
Comment (nicksanford) [reply services/datamanager/builtin/builtin.go:653]: This happens any time Reconfigure changes any of these config values: https://github.com/viamrobotics/rdk/pull/4164/files#diff-ffb86a3e2b364cc470f33937bb1ffb08ca2ef9d01b2734320daea7bee24a8943R621-R628

File: services/datamanager/builtin/builtin.go
Comment (nicksanford) [reply services/datamanager/builtin/builtin.go:692]: Yes it is the same however I don't see how I can dry this up without changing the behavior giv
en we need to propagate the initialization error in one case to terminate the loop, return no error in another case to continue the loop and continue the function in another case.

I'd also like to leave drying up to later PRs which are focused on reducing code complexity. I'm trying to keep this PR to the minimal change required to fix the current issues.

File: services/datamanager/builtin/builtin.go
Comment (nicksanford) [reply services/datamanager/builtin/builtin.go:704]: `syncConfigUpdated = true` means that the Reconfigure method was passed config which changed what data sync should do & that the thread that runs `propagateDataSyncConfig` has not yet handled it.

`reinitSyncer` means that `closeSyncer` should be called followed by `initSyncer`.

File: services/datamanager/builtin/builtin.go
Comment (nicksanford) [inline services/datamanager/builtin/builtin.go:154]: I added `syncerNeedsToBeReInitialized`, `syncConfigUpdated`, `closedCtx`, `closedCancelFn`, and `propagateDataSyncConfigWG`.

File: services/datamanager/builtin/builtin.go
Comment (nicksanford) [reply services/datamanager/builtin/builtin.go]: done

File: services/datamanager/builtin/builtin.go
Comment (nicksanford) [reply services/datamanager/builtin/builtin.go]: done

**Review by n0nick — APPROVED** [review]
LGTM but prefer to get another approval from Devin or Alexa, as I'm not familiar enough with the codebase

File: services/datamanager/builtin/builtin.go
Comment (agreenb) [reply services/datamanager/builtin/builtin.go:411]: What error would be returned if it was a local config? Wondering if we could instead add if errors.Is(err, cloud.ErrNotCloudManaged) or _whatever the error is when it's local_.

I ask this since we modeled the noop sync manager after the [noop package manager](https://github.com/viamrobotics/rdk/blob/a0b1b46751ae39cfa5b6ee60d460299360cb2f34/robot/packages/noop_package_manager.go#L10) that we use in RDK to download packages.

File: services/datamanager/builtin/builtin.go
Comment (agreenb) [inline services/datamanager/builtin/builtin.go:671]: Thanks for the decoupling of sync and capture configuration. Great improvement.

**Review by agreenb — APPROVED** [review]
Awesome! I'm going to approve to unblock, but I still have a question about the no-op sync manager. Either way, making the conditions explicit by disentangling the capture and sync configurations and handling if reconfiguration fails is necessary.

Thanks for your hard work on this! As always, would ask that you test the various scenarios: where reconfiguration would have failed before, where you have sync disabled then enable it then disable and see data flow through to app as expected, etc.

**Comment by agreenb** [comment]
Note: just read through your manual testing in the PR description. Perfect, thank you.

File: services/datamanager/builtin/builtin.go
Comment (nicksanford) [reply services/datamanager/builtin/builtin.go:411]: `ErrNotCloudManaged` is the error returned if it is a local config.
I think we are doing `if errors.Is(err, cloud.ErrNotCloudManaged) {` here: https://github.com/viamrobotics/rdk/pull/4164/files#diff-ffb86a3e2b364cc470f33937bb1ffb08ca2ef9d01b2734320daea7bee24a8943R681 which will terminate the goroutine trying to boot data capture: https://github.com/viamrobotics/rdk/pull/4164/files#diff-ffb86a3e2b364cc470f33937bb1ffb08ca2ef9d01b2734320daea7bee24a8943R659 as the viam-server using a local config means data sync will NEVER succeed for that viam server instance and as a result we should give up on trying to boot data captuer.

Does that seem reasonable?

Comment (nicksanford) [reply services/datamanager/builtin/builtin.go:411]: I don't understand what the noop package manager has to do with whether or not we give up on trying to boot data sync depending on if it is possible for viam-server to connect to the cloud.

**Review by vijayvuyyuru — APPROVED** [review]
Looks good to me, thanks for detangling all of this logic.

**Pull request merged**
